### PR TITLE
Added some Activation functions

### PR DESCRIPTION
**Pull request by AdarshKumar712**
Added the following activation functions along with tests

- [x] hardsigmoid
- [x] hardtanh 
- [x] thresholdrelu or trelu
- [x]  lisht 

- hardsigmoid and hardtanh are piecewise linear approximations and computationally effiecient than their counterparts, sigmoid and tanh respectively
- Threshold relu is used in Autoencoders as Threshold gated RELU
- LiSHT: Non-Parametric Linearly Scaled Hyperbolic Tangent Activation Function
  Ref: https://arxiv.org/pdf/1901.05894.pdf

### CODE REVIEW

File: test/activation.jl [hunk 1]
```diff
@@ -101,7 +113,7 @@ end
     end
 
     @testset "Test Integer64 and Integer32 inputs will force Float64 outputs" begin
-        test_value_int_input_forces_float64.(filter(x -> (x != relu && x != relu6), ACTIVATION_FUNCTIONS))
+        test_value_int_input_forces_float64.(filter(x -> (x != relu && x != relu6 ), ACTIVATION_FUNCTIONS))
```
Comment (staticfloat) [inline test/activation.jl]: I think this extra space is a mistake.

File: src/activation.jl [hunk 2]
```diff
@@ -17,6 +17,15 @@ const sigmoid = σ
   σ(x::ForwardDiff.Dual{T,Float32}) where T = σ_stable(x)
 end
 
+"""
+   hardσ(x) = max(0, min(1.0, 0.2 * x + 0.5))
+
+Segment-wise linear approximation of sigmoid
+Note: It should not be used with Regression tasks.
```
Comment (staticfloat) [inline src/activation.jl]: Can you give a reason why you've added this note?  Is there a paper that explains this?

File: src/activation.jl [hunk 3]
```diff
@@ -35,6 +44,14 @@ logσ(x::Real) = -softplus(-x)
 const logsigmoid = logσ
 
 
+"""
+    hardtanh(x) = max(-1, min(1, x))
+
+Segment-wise linear approximation of tanh. Cheaper  and  more  computational  efficient version of tanh.
+"""
+hardtanh(x::Real) = oftype(x / 1,max(-1, min( 1, x)))
```
Comment (staticfloat) [inline src/activation.jl]: Is there a paper that introduced `hardtanh()`?

File: test/activation.jl (hunk 1 above)
Comment (AdarshKumar712) [reply test/activation.jl]: Oh! Yes, That was left there by mistake.

File: src/activation.jl (hunk 3 above)
Comment (AdarshKumar712) [reply src/activation.jl]: I am not completely sure that if the following paper introduced this, but they have used this in NLP
http://www.jmlr.org/papers/volume12/collobert11a/collobert11a.pdf

Comment (AdarshKumar712) [reply src/activation.jl]: And they have claimed that this function is computationally cheaper than tanh itself.

File: src/activation.jl (hunk 2 above)
Comment (AdarshKumar712) [reply src/activation.jl]: I have added this note because the errors will be higher in case of Regression tasks while using piecewise approximations and also it has been observed to get better results on DL Binary classification tasks.

File: test/activation.jl (hunk 1 above)
Comment (AdarshKumar712) [reply test/activation.jl]: I will remove this

File: src/activation.jl (hunk 2 above)
Comment (AdarshKumar712) [reply src/activation.jl]: Another thing is that here I have implemented it the way it has been implemented in Tensorflow as the coefficient of x is 0.2. In original paper, it was taken as 0.5.(https://arxiv.org/pdf/1511.00363.pdf). Is that ok?

File: test/activation.jl [hunk 4]
```diff
@@ -1,6 +1,6 @@
 using NNlib, Test, Zygote
 
-ACTIVATION_FUNCTIONS = [σ, relu, leakyrelu, relu6, rrelu, elu, gelu, celu, swish, selu, softplus, softsign, logcosh, mish, tanhshrink, softshrink];
+ACTIVATION_FUNCTIONS = [σ,hardσ, hardtanh, relu, leakyrelu, relu6, rrelu, elu, gelu, celu, swish, lisht, selu, trelu, softplus, softsign, logcosh, mish, tanhshrink, softshrink];
```
Comment (matsueushi) [inline test/activation.jl]:
```suggestion
ACTIVATION_FUNCTIONS = [σ, hardσ, hardtanh, relu, leakyrelu, relu6, rrelu, elu, gelu, celu, swish, lisht, selu, trelu, softplus, softsign, logcosh, mish, tanhshrink, softshrink];
```
I think a space is missing

File: src/activation.jl (hunk 3 above)
Comment (matsueushi) [inline src/activation.jl]:
```suggestion
hardtanh(x::Real) = max(-one(x), min(one(x), x))
```
I think `hardtanh` doesn't have to force floating point outputs. Our `relu` and `relu6` return integers for integer inputs.

Comment (AdarshKumar712) [reply src/activation.jl]: Actually I am unsure about how to decide on that. Though without forcing floating point output, all tests are passing(making this as an exception for int to forced floating point test). So, I guess this doesn't have to force floating point outputs. Also, I think the same goes for trelu as well. If that is so, I will make changes in both. Thanks for pointing that out.
Other than unwanted computation cost, are there any other reasons why we ensure such typecastings?

Comment (matsueushi) [reply src/activation.jl]: I'm not sure the background behind it, but this test 
https://github.com/FluxML/NNlib.jl/blob/c82a76d40cf5c34e4a1eed2655aa30b464e3e408/test/activation.jl#L106-L110

was added in the commit https://github.com/FluxML/NNlib.jl/commit/b698441cf0dfbf08875fa3364b44d95bad4ba104 by @staticfloat .

File: src/activation.jl (hunk 2 above)
Comment (staticfloat) [reply src/activation.jl]: It should probably be a parameter with a default value (the TF value is fine, I think).

File: src/activation.jl (hunk 3 above)
Comment (staticfloat) [reply src/activation.jl]: Oh, it's definitely cheaper, I can tell you that.  ;)

Reading that paper, they state that the `HardTanh` was first introduced in `(Collobert, 2004)` which turns out to be the same author's thesis: http://ronan.collobert.org/pub/matos/2004_phdthesis_lip6.pdf

So we'll just cite that as the originating paper.  :)

Comment (staticfloat) [reply src/activation.jl]: Unless we have a reason to force floating-point (e.g. for gradients) I prefer to keep datatypes as they are.

File: src/activation.jl (hunk 2 above)
Comment (AdarshKumar712) [reply src/activation.jl]: Ok, I will change it to a parameter

File: src/activation.jl (hunk 3 above)
Comment (AdarshKumar712) [reply src/activation.jl]: Ok, I will add that with the function

Comment (AdarshKumar712) [reply src/activation.jl]: I will the unwanted typecasting then

File: test/activation.jl (hunk 4 above)
Comment (AdarshKumar712) [reply test/activation.jl]: I will add that

**Comment by AdarshKumar712** [comment]
@staticfloat @matsueushi I have made the changes. Please have a look.

File: src/activation.jl (hunk 2 above)
Comment (staticfloat) [reply src/activation.jl]: I don't think we should include that note, at least not without a publication explaining why.  I don't think we have enough proof to put that strong of an assertion into the source code.

File: test/activation.jl [hunk 5]
```diff
@@ -114,7 +126,18 @@ end
             @test typeof(relu6(Int64(1))) == Int64
             @test typeof(relu6(Int32(1))) == Int32
         end
-
+        
+        @testset "hardtanh: " begin
+            # relu6 doesn't have to force floating point outputs
```
Comment (matsueushi) [inline test/activation.jl]:
```suggestion
            # hardtanh doesn't have to force floating point outputs
```

File: test/activation.jl [hunk 6]
```diff
@@ -114,7 +126,18 @@ end
...
+            @test typeof(hardtanh(Int64(1))) == Int64
+            @test typeof(hardtanh(Int32(1))) == Int32
+        end
+        
+        @testset "trelu: " begin
+            # relu6 doesn't have to force floating point outputs
```
Comment (matsueushi) [inline test/activation.jl]:
```suggestion
            # trelu doesn't have to force floating point outputs
```

File: src/activation.jl [hunk 7]
```diff
@@ -136,6 +166,19 @@ function celu(x::Real, α::Real = one(x))
...
+
+"""
+    trelu(x, theta = 1.0) = x > theta ? x : 0 
+
+Threshold Gated Rectified Linear   
+See [ThresholdRelu](https://arxiv.org/pdf/1402.3337.pdf)
+"""
+function trelu(x::Real,theta = one(x)) 
+    return ifelse(x> theta, x/ one(x), zero(x))
+end
```
Comment (matsueushi) [inline src/activation.jl]: Since we don't need to force floating point outputs, we can write just as
```julia
trelu(x::Real, theta=one(x)) = ifelse(x > theta, x, zero(x))
```

File: src/activation.jl (hunk 2 above)
Comment (AdarshKumar712) [reply src/activation.jl]: Ok, I will remove that

**Comment by AdarshKumar712** [comment]
Made all the changes asked above

**Comment by matsueushi** [comment]
Looks good to me!

File: src/activation.jl [hunk 8]
```diff
@@ -17,6 +17,15 @@ const sigmoid = σ
   σ(x::ForwardDiff.Dual{T,Float32}) where T = σ_stable(x)
 end
 
+"""
+   hardσ(x, a=0.2) = max(0, min(1.0, a * x + 0.5))
+
+Segment-wise linear approximation of sigmoid
+See: [BinaryConnect: Training Deep Neural Networks withbinary weights during propagations](https://arxiv.org/pdf/1511.00363.pdf)
+"""
+hardσ(x::Real, a=0.2) = oftype(x/1, max(zero(x/1), min(one(x/1), oftype(x/1,a) * x + oftype(x/1,0.5))))
```
Comment (CarloLucibello) [inline src/activation.jl]: I think division by `1` should be replaced by division by `one(x)` like fo other activations

Comment (AdarshKumar712) [reply src/activation.jl]: Actually one(x) or 1 doesn't make much difference here. In both the cases output precision remains the same. Moreover, division by is faster than division by one(x)
```
julia> x = Float32(1)
1.0f0

julia> @time x/one(x)
  0.002585 seconds (28 allocations: 1.875 KiB)
1.0f0

julia> @time x/1
  0.000009 seconds (5 allocations: 176 bytes)
1.0f0
```
